src/back/spv/writer.rs

```diff
@@ -732,61 +704,93 @@ impl Writer {
     /// Primitive Instructions
     ///
 
-    fn parse_type_declaration(
+    fn write_scalar(&self, id: Word, kind: crate::ScalarKind, width: u8) -> Instruction {
+        let bits = (width * BITS_PER_BYTE) as u32;
+        match kind {
+            crate::ScalarKind::Sint => self.instruction_type_int(id, bits, Signedness::Signed),
+            crate::ScalarKind::Uint => self.instruction_type_int(id, bits, Signedness::Unsigned),
+            crate::ScalarKind::Float => self.instruction_type_float(id, bits),
+            crate::ScalarKind::Bool => self.instruction_type_bool(id),
+        }
+    }
+
+    fn parse_to_spirv_storage_class(&self, class: crate::StorageClass) -> spirv::StorageClass {
+        match class {
+            crate::StorageClass::Constant => spirv::StorageClass::UniformConstant,
+            crate::StorageClass::Function => spirv::StorageClass::Function,
+            crate::StorageClass::Input => spirv::StorageClass::Input,
+            crate::StorageClass::Output => spirv::StorageClass::Output,
+            crate::StorageClass::Private => spirv::StorageClass::Private,
+            crate::StorageClass::StorageBuffer => spirv::StorageClass::StorageBuffer,
+            crate::StorageClass::Uniform => spirv::StorageClass::Uniform,
+            crate::StorageClass::WorkGroup => spirv::StorageClass::Workgroup,
+        }
+    }
+
+    fn write_type_declaration_local(
+        &mut self,
+        arena: &crate::Arena<crate::Type>,
+        local_ty: LocalType,
+    ) -> Word {
+        let id = self.generate_id();
+        let instruction = match local_ty {
+            LocalType::Scalar { kind, width } => self.write_scalar(id, kind, width),
+            LocalType::Vector { size, .. } => {
+                let scalar_id = self.get_type_id(arena, LookupType::Local(local_ty));
+                self.instruction_type_vector(id, scalar_id, size)
+            }
+            LocalType::Pointer { .. } => unimplemented!(),
+        };
+
+        self.lookup_type.insert(LookupType::Local(local_ty), id);
```

Comment:
I see what you did here. Let's cut the loop. Instead of doing `match (entry) -> call function -> insert entry` just do `match (entry) -> insert entry` as we have the `Entry::Vacant` there.